### PR TITLE
ts-730 add reading token from commit-log/bookkeeper

### PR DESCRIPTION
**Pull request by tomskikh**
#730

### CODE REVIEW

File: tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ManyClientsSingleNodeServerInterconnectionTest.scala [hunk 1]
```diff
@@ -287,7 +287,7 @@ class ManyClientsSingleNodeServerInterconnectionTest
     }
   }
 
-  "One client" should "put producer transaction: Opened; the second one: Invalid. The Result is opened transaction." in {
+  "One client" should "put producer transaction: Opened; the second one: Invalid. The Result is invalid transaction." in {
```
Comment (bwsw) [inline tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ManyClientsSingleNodeServerInterconnectionTest.scala]: This test should be removed, It checks very strange behaviour. To discuss @tomskikh @placccebo

File: tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ServerClientInterconnectionTest.scala [hunk 2]
```diff
@@ -315,7 +315,7 @@ class ServerClientInterconnectionTest
     val transactionServer = bundle.transactionServer
```
Comment (bwsw) [inline tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ServerClientInterconnectionTest.scala]: @placccebo @tomskikh 
I suppose all Interconnection (which means Reconnection) should be removed
To discuss first.

File: tstreams-transaction-server/src/main/scala/com/bwsw/commitlog/CommitLog.scala [hunk 3]
```diff
@@ -173,4 +181,14 @@ class CommitLog(seconds: Int,
...
+}
+
+
+object CommitLog {
+  def intToBytes(i: Int): Array[Byte] =
+    ByteBuffer.allocate(Integer.BYTES).putInt(i).array()
+
+  def bytesToInt(bytes: Array[Byte]): Int =
+    ByteBuffer.wrap(bytes).getInt()
 }
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/commitlog/CommitLog.scala]: add a new line in the end of file.

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/TestSingleNodeServer.scala [hunk 4]
```diff
@@ -51,17 +51,17 @@ class TestSingleNodeServer(authenticationOpts: AuthenticationOptions,
     subscribersUpdateOptions,
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/TestSingleNodeServer.scala]: rename class to SingleNodeTestingServer

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala [hunk 5]
```diff
@@ -24,6 +24,7 @@ import com.bwsw.tstreamstransactionserver.netty.server.commitLogService.Schedule
 import com.bwsw.tstreamstransactionserver.netty.server.handler.PredefinedContextHandler
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala]: Rename to ConsumerCheckpointPutHandler

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/data/PutProducerStateWithDataHandler.scala [hunk 6]
```diff
@@ -23,14 +23,13 @@ import com.bwsw.tstreamstransactionserver.netty.server.TransactionServer
 import com.bwsw.tstreamstransactionserver.netty.server.batch.Frame
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/data/PutProducerStateWithDataHandler.scala]: rename to ProducerStateWithDataPutHandler

Comment (bwsw) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/data/PutProducerStateWithDataHandler.scala]: rename other handlers according to the scheme.

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionHandler.scala [hunk 7]
```diff
@@ -58,12 +58,12 @@ class PutTransactionHandler(server: TransactionServer,
   }
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionHandler.scala]: rename to TransactionPutHandler

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionsHandler.scala [hunk 8]
```diff
@@ -22,13 +22,13 @@ import com.bwsw.tstreamstransactionserver.netty.server.TransactionServer
 import com.bwsw.tstreamstransactionserver.netty.server.batch.Frame
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionsHandler.scala]: rename to TransactionsPutHandler

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/transactionMetadataService/test/TestProducerTransactionsCleaner.scala [hunk 9]
```diff
@@ -20,13 +20,15 @@
 package com.bwsw.tstreamstransactionserver.netty.server.transactionMetadataService.test
 
 import com.bwsw.tstreamstransactionserver.netty.server.Notifier
+import com.bwsw.tstreamstransactionserver.netty.server.authService.OpenedTransactions
 import com.bwsw.tstreamstransactionserver.netty.server.db.KeyValueDbManager
 import com.bwsw.tstreamstransactionserver.netty.server.transactionMetadataService.{ProducerTransactionRecord, ProducerTransactionsCleaner}
 import com.bwsw.tstreamstransactionserver.rpc.ProducerTransaction
 
 class TestProducerTransactionsCleaner(rocksDB: KeyValueDbManager,
```
Comment (bwsw) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/transactionMetadataService/test/TestProducerTransactionsCleaner.scala]: Rename to ProducerTransactionsTestingCleaner, rename package to testing.

File: tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala [hunk 10]
```diff
@@ -19,15 +19,16 @@
...
 package com.bwsw.commitlog
 
+import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}
 import org.scalatest.{FlatSpec, Matchers}
 
 class CommitLogRecordTest extends FlatSpec with Matchers {
 
   "CommitLogRecord" should "be serialized/deserialized" in {
-    val record1 = new CommitLogRecord(0: Byte, "test_data".getBytes())
+    val record1 = CommitLogRecord(0, System.currentTimeMillis(), 12345, "test_data".getBytes())
```
Comment (bwsw) [inline tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala]: move magic numbers to vars.

File: tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala [hunk 11]
```diff
@@ -19,15 +19,16 @@
...
     CommitLogRecord.fromByteArray(record1.toByteArray).right.get shouldBe record1
 
-    val record2 = new CommitLogRecord(-5: Byte, "".getBytes())
+    val record2 = CommitLogRecord(-5, System.currentTimeMillis(), 762234, "".getBytes())
```
Comment (bwsw) [inline tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala]: move magic numbers to vars.

File: tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogTest.scala [hunk 12]
```diff
@@ -24,94 +24,98 @@ import java.nio.file._
...
 
   "CommitLog" should "write correctly (OnRotation policy)" in {
     val cl = new CommitLog(1, directory, OnRotation, fileIDGen)
-    val f1 = cl.putRec(record, 0)
+    val f1 = cl.putRec(record, 0, token)
     val fileF1 = new File(f1)
     fileF1.exists() shouldBe true
-    fileF1.length() == 0 shouldBe true
+    fileF1.length() shouldBe 0
     Thread.sleep(1100)
```
Comment (bwsw) [inline tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogTest.scala]: move to var, use calculation to provide the meaning how it happens that 1100 is OK.

File: tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/multiNode/bookkeperService/hierarchy/BookkeeperToRocksWriterTest.scala [hunk 13]
```diff
@@ -313,14 +319,15 @@ class BookkeeperToRocksWriterTest
...
-    val firstTreeRecords =
+    val firstTreeRecords = new Record(Frame.TokenCreatedType, firstTimestamp, token, Array.emptyByteArray) +:
       genProducerTransactionsWrappedInRecords(
         atomicLong,
         producerTransactionsNumber,
         stream.id,
         partition,
         Opened,
-        50000L
+        50000L,
```
[the remark this answers is not in the document]
Comment (tomskikh) [reply tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/multiNode/bookkeperService/hierarchy/BookkeeperToRocksWriterTest.scala]: Done

File: tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/multiNode/bookkeperService/hierarchy/BookkeeperToRocksWriterTest.scala [hunk 14]
```diff
@@ -337,10 +344,10 @@ class BookkeeperToRocksWriterTest
         stream.id,
         partition,
         Checkpointed,
-        50000L
+        50000L,
```
[the remark this answers is not in the document]
Comment (tomskikh) [reply tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/multiNode/bookkeperService/hierarchy/BookkeeperToRocksWriterTest.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/commitlog/CommitLog.scala (hunk 3 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/commitlog/CommitLog.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/TestSingleNodeServer.scala (hunk 4 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/TestSingleNodeServer.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala (hunk 5 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionsHandler.scala (hunk 8 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionsHandler.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionHandler.scala (hunk 7 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/metadata/PutTransactionHandler.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/data/PutProducerStateWithDataHandler.scala (hunk 6 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/data/PutProducerStateWithDataHandler.scala]: Done

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/transactionMetadataService/test/TestProducerTransactionsCleaner.scala (hunk 9 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/transactionMetadataService/test/TestProducerTransactionsCleaner.scala]: Done

File: tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala (hunk 10 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala]: Done

File: tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala (hunk 11 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogRecordTest.scala]: Done

File: tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogTest.scala (hunk 12 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/test/scala/com/bwsw/commitlog/CommitLogTest.scala]: Done

File: tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ServerClientInterconnectionTest.scala (hunk 2 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ServerClientInterconnectionTest.scala]: We don't have tests for reconnection. ServerClientInterconnectionTest contains tests to check that data client send data if it is connected to server or disconnects and throw exception if master isn't available or changed. I suggest to split and rename that tests.

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala (hunk 5 above)
Comment (placccebo) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala]: This method isn't used at all, to discuss.

File: tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ManyClientsSingleNodeServerInterconnectionTest.scala (hunk 1 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/it/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/ManyClientsSingleNodeServerInterconnectionTest.scala]: I looked [ProducerTransactionStateMachine](https://github.com/bwsw/t-streams/blob/cb2ef5b5c063cb1f4d6f0136e51c5900ea4ee6be/tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/transactionMetadataService/stateHandler/ProducerTransactionStateMachine.scala#L61) and [ProducerTransactionState](https://github.com/bwsw/t-streams/blob/cb2ef5b5c063cb1f4d6f0136e51c5900ea4ee6be/tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/transactionMetadataService/stateHandler/ProducerTransactionState.scala). Server ignores producer transactions with state "invalid" sent by client. That test should be reverted, it checks that server doesn't invalidates transaction if one client opens transaction and other client sends "invalid" state for it.

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+/** Contains opened transactions
+  *
+  * @param tokenCache cache for tokens
+  * @author Pavel Tomskikh
+  */
+class OpenedTransactions(val tokenCache: TokenCache) {
+
+  private val transactions = TrieMap.empty[Long, Int]
+
+  /** Adds transaction with specific token if it absent and token is valid
```
Comment (placccebo) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala]: it is absent

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+  /** Removes transaction
+    *
+    * @param transaction transaction ID
+    */
+  def remove(transaction: Long): Unit = transactions -= transaction
+
+  /** Checks that transaction is valid
+    *
+    * @param transaction transaction ID
+    * @return true, if transaction is present and it has valid token, false otherwise
```
Comment (placccebo) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala]: be + V3 for passive voice - fix a mistake

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/TokenCache.scala [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+package com.bwsw.tstreamstransactionserver.netty.server.authService
+
+import java.util.concurrent.TimeUnit
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.duration.TimeUnit
+
+/** Contains valid tokens with it's creation time. Used to handle tokens on a slave.
```
Comment (placccebo) [inline tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/TokenCache.scala]: it's -> their

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala (hunk 15 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala]: Fixed

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala (hunk 16 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/OpenedTransactions.scala]: Fixed

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/TokenCache.scala (hunk 17 above)
Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/authService/TokenCache.scala]: Fixed

File: tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala (hunk 5 above)
Comment (bwsw) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala]: @tomskikh @placccebo 
remove that unused method.

Comment (bwsw) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala]: @tomskikh @placccebo
remove that unused method.

Comment (tomskikh) [reply tstreams-transaction-server/src/main/scala/com/bwsw/tstreamstransactionserver/netty/server/singleNode/hanlder/consumer/PutConsumerCheckpointHandler.scala]: Removed

**Pull request merged**